homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,187 @@
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SOMFY_COMPONENTS = ['cover']
+
+
+def setup(hass, config):
+    """Set up the Somfy component."""
+    from pymfy.api.somfy_api import SomfyApi
+
+    hass.data[DOMAIN] = {}
+    is_ready = False
+
+    # This is called to create the redirect so the user can Authorize Home .
+    redirect_uri = '{}{}'.format(
+        hass.config.api.base_url, SOMFY_AUTH_CALLBACK_PATH)
+    conf = config[DOMAIN]
+    api = SomfyApi(conf.get(CONF_CLIENT_ID),
+                   conf.get(CONF_CLIENT_SECRET),
+                   redirect_uri, hass.config.path(DEFAULT_CACHE_PATH))
+    hass.data[DOMAIN][API] = api
+
+    if not api.token:
+        authorization_url, _ = api.get_authorization_url()
+        hass.components.persistent_notification.create(
+            'In order to authorize Home Assistant to view your Somfy devices'
+            ' you must visit this <a href="{}" target="_blank">link</a>.'
+            .format(authorization_url),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID
+        )
+        hass.http.register_view(SomfyAuthCallbackView(config))
+        is_ready = True
+    else:
+        if update_all_devices(hass):
+            is_ready = True
+            for component in SOMFY_COMPONENTS:
+                discovery.load_platform(hass, component, DOMAIN, {}, config)
+
+    return is_ready
+
+
+class SomfyAuthCallbackView(HomeAssistantView):
+    """Handle OAuth finish callback requests."""
+
+    url = SOMFY_AUTH_CALLBACK_PATH
+    name = 'auth:somfy:callback'
+    requires_auth = False
+
+    def __init__(self, config):
+        """Initialize the OAuth callback view."""
+        self.config = config
+
+    @callback
+    def get(self, request):
+        """Finish OAuth callback request."""
+        from aiohttp import web
+        from oauthlib.oauth2 import MismatchingStateError
+        from oauthlib.oauth2 import InsecureTransportError
+
+        hass = request.app['hass']
+        response = web.HTTPFound('/')
+
+        try:
+            code = request.query.get('code')
+            hass.data[DOMAIN][API].request_token(code=code)
+            hass.async_add_job(setup, hass, self.config)
+            hass.components.persistent_notification.dismiss(NOTIFICATION_ID)
+            hass.components.persistent_notification.create(
+                "Somfy has been successfully authorized!",
+                title=NOTIFICATION_TITLE,
+                notification_id=NOTIFICATION_ID
+            )
+        except MismatchingStateError:
+            _LOGGER.error("OAuth state not equal in request and response.",
+                          exc_info=True)
+        except InsecureTransportError:
+            _LOGGER.error("Somfy redirect URI %s is insecure.", request.url,
+                          exc_info=True)
+
+        return response
+
+
+class SomfyEntity(Entity):
+    """Representation of a generic Somfy device."""
+
+    def __init__(self, device, hass):
```

Comment:
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.